cradle-core/src/main/java/com/exactpro/cradle/BookInfo.java

```diff
@@ -160,12 +165,99 @@ public PageInfo getPage(PageId pageId)
+									   @Nonnull Order order) {
+		Instant leftTimestamp = calculateBound(leftBoundTimestamp, this::getFirstPage);
+		if (leftTimestamp == null) {
+			return emptyIterator();
+		}
+		Instant rightTimestamp = calculateBound(rightBoundTimestamp, this::getLastPage);
+		if (rightTimestamp == null) {
+			return emptyIterator();
+		}
+
+		if (leftTimestamp.isAfter(rightTimestamp)) {
+			throw new IllegalArgumentException(
+					"Left bound '" + leftBoundTimestamp + " -> " + leftTimestamp +
+							"' must be <= right bound '" + rightBoundTimestamp + " -> " + rightTimestamp + "'"
+			);
+		}
+		long leftBoundEpochDay = getEpochDay(leftTimestamp);
+		long rightBoundEpochDay = getEpochDay(rightTimestamp);
+
+		Stream.Builder<Long> builder = Stream.builder();
+		switch (order) {
+            case DIRECT:
+				for (long i = leftBoundEpochDay; i <= rightBoundEpochDay; i++) {
+					builder.add(i);
+				}
+				break;
+            case REVERSE:
+				for (long i = rightBoundEpochDay; i >= leftBoundEpochDay; i--) {
+					builder.add(i);
+				}
+                break;
+            default:
+                throw new IllegalStateException("Unexpected value: " + order);
+        }
+		return builder.build()
+				.flatMap(epochDay -> getPageInterval(epochDay).getPages(leftTimestamp, rightTimestamp, order))
+				.iterator();
+	}
+
+	public PageInfo findPage(Instant timestamp) {
+		// Search in the page interval related to the timestamp
+		long epochDay = getEpochDay(timestamp);
+		IPageInterval pageInterval = getPageInterval(epochDay);
+		PageInfo pageInfo = pageInterval.find(timestamp);
 		METRICS.incRequest(id.getName(), pageInterval.getCacheName(), FIND);
-		return pageInterval.find(timestamp);
+		if (pageInfo != null) {
+			return pageInfo;
+		}
+
+		// Search in page intervals before the timestamp
+		PageInfo firstPage = getFirstPage();
+		if (firstPage == null) {
+			return null;
+		}
+		if (firstPage.getStarted().isAfter(timestamp)) {
+			return null;
+		}
+
+		long firstEpochDay = getEpochDay(firstPage.getStarted());
+		if (epochDay < firstEpochDay) {
+			return null;
+		} else if (epochDay > firstEpochDay) {
+			for (long previousEpochDay = epochDay - 1; previousEpochDay >= firstEpochDay; previousEpochDay--) {
+				IPageInterval previousPageInterval = getPageInterval(previousEpochDay);
+				if (previousPageInterval.size() == 0) {
+					// page interval without new pages
+					continue;
+				}
+				PageInfo previousPageInfo = previousPageInterval.find(timestamp);
+				METRICS.incRequest(id.getName(), previousPageInterval.getCacheName(), FIND);
+				if (previousPageInfo == null) {
+					throw new IllegalStateException(
+							"Previous page interval isn't empty but doesn't contain page covers date " + timestamp
+					);
+				}
+				return previousPageInfo;
+			}
+			throw new IllegalStateException(
+					"First page is before requested timestamp, but neither cache doesn't contain appropriate page, requested: " +
```

Comment:
There is no double negation in English. This is one of the options how you can correct the sentence.
```suggestion
					"First page is before requested timestamp, but neither cache contains the appropriate page, requested: " +
```